### PR TITLE
mm-241 - adds aria labels to right hand side elements 

### PR DESCRIPTION
**Pull request by maisnamrajusingh**
#### Summary
Hovering over the elements on the right hand side should display the labels for those items. 
#### Ticket Link
https://github.com/mattermost/mattermost-plugin-github/issues/241

### CODE REVIEW

**Review by larkox — APPROVED** [review]
LGTM

**Review by hanzei — APPROVED** [review]
LGTM :+1:

**Review by hanzei — COMMENTED** [review]
Should the PR/Issue title also have a label?

**Comment by hanzei** [comment]
And what about the milestone icon?

**Comment by jasonblais** [comment]
@hanzei is this ready to merge?  Looks like it has two approvals. We can create follow-up tickets for the additional labels and icons?

File: webapp/src/components/sidebar_right/github_items.tsx [hunk 1]
```diff
@@ -185,13 +188,28 @@ function GithubItems(props: GithubItemsProps) {
         if (item.status) {
             switch (item.status) {
             case 'success':
-                status = (<span style={{...style.icon, ...style.iconSucess}}><TickIcon/></span>);
+                status = (
+                    <span
+                        title={'Success'}
+                        style={{...style.icon, ...style.iconSucess}}
+                    ><TickIcon/></span>
```
Comment (mickmister) [inline webapp/src/components/sidebar_right/github_items.tsx:195]: Can we format the JSX for this block and the two blocks below like this?

```suggestion
                    <span
                        title={'Success'}
                        style={{...style.icon, ...style.iconSucess}}
                    >
                        <TickIcon/>
                    </span>
```

**Review by mickmister — CHANGES_REQUESTED** [review]
Thanks @maisnamrajusingh, I have a request to reformat some JSX indenting. Also, I agree with @hanzei that we should add the aria-labels to the PR title and milestones, and I would like to add them to the PR/issue labels as well.

File: webapp/src/components/sidebar_right/github_items.tsx (hunk 1 above)
Comment (mickmister) [reply webapp/src/components/sidebar_right/github_items.tsx]: @maisnamrajusingh By "and the two blocks below" I meant to apply this change to the two blocks below this block as well. Can you please apply the change to the other two `span` blocks?

**Review by hanzei — CHANGES_REQUESTED** [review]
Blocking on by questions above: https://github.com/mattermost/mattermost-plugin-github/pull/462#pullrequestreview-720128972

**Comment by maisnamrajusingh** [comment]
@hanzei  I have assigned the changes to @sanjaydemansol He will take a look into it

**Comment by mickmister** [comment]
I'm also blocking on https://github.com/mattermost/mattermost-plugin-github/pull/462#discussion_r685672062

File: webapp/src/components/sidebar_right/github_items.tsx (hunk 1 above)
Comment (sanjaydemansol) [reply webapp/src/components/sidebar_right/github_items.tsx]: sure, done.

**Comment by maisnamraju** [comment]
> Should the PR/Issue title also have a label?

I guess it should have one, @sanjaydemansol  can you add it.

**Comment by sanjaydemansol** [comment]
> Should the PR/Issue title also have a label?

Please elaborate, what should be the content of label?

File: webapp/src/components/sidebar_right/github_items.tsx [hunk 2]
```diff
@@ -5,10 +5,10 @@ import * as React from 'react';
 
 import * as CSS from 'csstype';
 
-import {Badge, Tooltip, OverlayTrigger} from 'react-bootstrap';
+import {Badge, OverlayTrigger, Tooltip} from 'react-bootstrap';
 import {Theme} from 'mattermost-redux/types/preferences';
-import {makeStyleFromTheme, changeOpacity} from 'mattermost-redux/utils/theme_utils';
-import {GitPullRequestIcon, IssueOpenedIcon, IconProps} from '@primer/octicons-react';
+import {changeOpacity, makeStyleFromTheme} from 'mattermost-redux/utils/theme_utils';
+import {GitPullRequestIcon, IconProps, IssueOpenedIcon} from '@primer/octicons-react';
```
Comment (mickmister) [inline webapp/src/components/sidebar_right/github_items.tsx:11]: None of these changes seem to be necessary for this PR. Can you please double check your changes when submitting the PRs to ensure all changes are necessary for the ticket?

**Review by mickmister — APPROVED** [review]
LGTM, one non-blocking request to remove changes that are unrelated to the PR

File: webapp/src/components/sidebar_right/github_items.tsx (hunk 2 above)
Comment (mohit-deshwal) [reply webapp/src/components/sidebar_right/github_items.tsx:11]: They are probably changed by ESLint, since imports are same. Just sequence is changed. do you want me to revert this part?

Comment (sanjaydemansol) [reply webapp/src/components/sidebar_right/github_items.tsx:11]: They are probably changed by ESLint, since imports are same. Just sequence is changed. do you want me to revert this part?

**Review by hanzei — COMMENTED** [review]
Did do add a label for the milestone icon?

**Comment by sanjaydemansol** [comment]
> Did do add a label for the milestone icon?

What should be the content of label? "milestone" ?

**Comment by hanzei** [comment]
The milestone name should also be part of the label.

**Comment by sanjaydemansol** [comment]
> The milestone name should also be part of the label.

Used word "Milestone", 2/5..its more helpful to know what `1.2.3` actually is. Let me know If any changes needed. @hanzei thanks.

File: webapp/src/components/sidebar_right/github_items.tsx [hunk 3]
```diff
@@ -165,6 +168,7 @@ function GithubItems(props: GithubItemsProps) {
         if (item.milestone) {
             milestone = (
                 <span
+                    title={'Milestone'}
```
Comment (hanzei) [inline webapp/src/components/sidebar_right/github_items.tsx:171]: Well, the point of screen reader support is to present all needed information for the screen reader. Just knowing _that_ an Issue/PR has a Milestone is not as useful as knowing to _which_ Milestone it's assigned to.

File: webapp/src/components/sidebar_right/github_items.tsx [hunk 4]
```diff
@@ -185,13 +189,34 @@ function GithubItems(props: GithubItemsProps) {
         if (item.status) {
             switch (item.status) {
             case 'success':
-                status = (<span style={{...style.icon, ...style.iconSucess}}><TickIcon/></span>);
+                status = (
+                    <span
+                        title={'Success'}
```
Comment (hanzei) [inline webapp/src/components/sidebar_right/github_items.tsx:194]: Can this be more descriptive? Maybe use these same wording as GitHub

**Review by hanzei — DISMISSED** [review]
When testing with https://chrome.google.com/webstore/detail/screen-reader/kgejglhpjiefppelpmljglcjbhoiplfn?hl=en the following elements were not read out:
- CI check
- Labels
- Milestone
- Number of reviews
- Time when the PR/issue was opened

Could you please check why they are not read out?

File: webapp/src/components/sidebar_right/github_items.tsx (hunk 3 above)
Comment (sanjaydemansol) [reply webapp/src/components/sidebar_right/github_items.tsx:171]: > I've noticed that the different elements/icons in the RHS PR list view do not have any aria-labels, so hovering over them has no effect. Some examples can be seen in this screenshot, particularly the PR link and the build status icon
@hanzei based on issue Author's comment, seems this PR should make hovering over icons more descriptive. Like what an icon signifies. Your opinion?

File: webapp/src/components/sidebar_right/github_items.tsx (hunk 4 above)
Comment (sanjaydemansol) [reply webapp/src/components/sidebar_right/github_items.tsx:194]: depends on reply to https://github.com/mattermost/mattermost-plugin-github/pull/462#discussion_r747668878

File: webapp/src/components/sidebar_right/github_items.tsx (hunk 3 above)
Comment (hanzei) [reply webapp/src/components/sidebar_right/github_items.tsx:171]: Yes, the hover information should be descriptive.

**Comment by hanzei** [comment]
@maisnamrajusingh Did you noticed by comment at https://github.com/mattermost/mattermost-plugin-github/pull/462#pullrequestreview-803526431?

**Comment by sanjaydemansol** [comment]
> @maisnamrajusingh Did you noticed by comment at [#462 (review)](https://github.com/mattermost/mattermost-plugin-github/pull/462#pullrequestreview-803526431)?

@hanzei here is suggestion from author https://github.com/mattermost/mattermost-plugin-github/issues/241#issuecomment-885384763 for similar request https://github.com/mattermost/mattermost-plugin-github/issues/241#issuecomment-884797740. 
I guess  https://chrome.google.com/webstore/detail/screen-reader/kgejglhpjiefppelpmljglcjbhoiplfn?hl=en is not applicable in our case, since this PR is to make Icons user friendly instead of adding Accessibility.

**Comment by mickmister** [comment]
@sanjaydemansol @hanzei The original intention of the ticket was to add accessibility, though I got confused and forgot that the browser's native tooltip from the `title` attribute also aids in accessbility.

I propose we introduce the `aria-label`s as originally planned, as well as using the tooltip functionality that's used throughout the rest of the application, using the `OverlayTrigger` and `Tooltip` components from `react-bootstrap`:

https://github.com/mattermost/mattermost-webapp/blob/242a18c68908bbcfee0effd5bc6804d711670c4a/components/channel_header/components/header_icon_wrapper.tsx#L134

**Comment by sanjaydemansol** [comment]
> @sanjaydemansol @hanzei The original intention of the ticket was to add accessibility, though I got confused and forgot that the browser's native tooltip from the `title` attribute also aids in accessbility.
> 
> I propose we introduce the `aria-label`s as originally planned, as well as using the tooltip functionality that's used throughout the rest of the application, using the `OverlayTrigger` and `Tooltip` components from `react-bootstrap`:
> 
> https://github.com/mattermost/mattermost-webapp/blob/242a18c68908bbcfee0effd5bc6804d711670c4a/components/channel_header/components/header_icon_wrapper.tsx#L134

@mickmister thank you for the clarity. Will make changes to the PR.

**Review by sibasankarnayak — APPROVED** [review]
LGTM

**Comment by sanjaydemansol** [comment]
@mickmister @hanzei, is this what we are looking for? (latest commit). Thanks.

File: webapp/src/components/sidebar_right/github_items.tsx [hunk 5]
```diff
@@ -186,13 +190,54 @@ function GithubItems(props: GithubItemsProps) {
...
+                        }
+                    >
+                        <span
+                            aria-label={'Pending'}
+                            style={{...style.icon, ...style.iconPending}}
+                        >
+                            <DotIcon/>
+                        </span>
+                    </OverlayTrigger>
+                );
```
Comment (hanzei) [inline webapp/src/components/sidebar_right/github_items.tsx:220]: This looks good. :+1:

**Comment by maisnamrajusingh** [comment]
@hanzei I'm taking over this one. Just confirming if only the comment below is the remaining fix needed on this one ?


> When testing with https://chrome.google.com/webstore/detail/screen-reader/kgejglhpjiefppelpmljglcjbhoiplfn?hl=en the following elements were not read out:
> 
> * CI check
> * Labels
> * Milestone
> * Number of reviews
> * Time when the PR/issue was opened
> 
> Could you please check why they are not read out?

**Comment by hanzei** [comment]
@maisnamrajusingh That is indeed the case.

File: webapp/src/components/sidebar_right/github_items.tsx (hunk 2 above)
Comment (maisnamrajusingh) [reply webapp/src/components/sidebar_right/github_items.tsx]: fixed with https://github.com/mattermost/mattermost-plugin-github/pull/462#:~:text=commit%20by%20sanjay-,443aa19,-Merge%20state

**Review by maisnamrajusingh — COMMENTED** [review]
@hanzei this field has a lot of text, if we were to insert it into an aria-label would it make sense to do put it into a variable and reuse the variable ? https://github.com/mattermost/mattermost-plugin-github/blob/b6c516c0ef168f3360cbdb98e9cab1792b6a519c/webapp/src/components/sidebar_right/github_items.tsx#L290-L308
```
             
                <div
                    className='light'
                    style={style.subtitle}
                >
                    {item.created_at && ('Opened ' + formatTimeSince(item.created_at) + ' ago')}
                    {userName && ' by ' + userName}
                    {(item.created_at || userName) && '.'}
                    {milestone}
                    {item.reason ? (<>
                        {(item.created_at || userName || milestone) && (<br/>)}
                        {item.updated_at && (formatTimeSince(item.updated_at) + ' ago')}{<br/>}
                        {notificationReasons[item.reason]}
                    </>) : null }
                </div>
              ```

File: webapp/package-lock.json [hunk 6]
```diff
@@ -17163,4 +17163,4 @@
       "dev": true
     }
   }
-}
```
Comment (mickmister) [inline webapp/package-lock.json:17166]: Please revert this change to keep the commit history of this file clean

File: webapp/src/components/sidebar_right/github_items.tsx [hunk 7]
```diff
@@ -381,11 +443,18 @@ function getReviewText(item: Item, style: any, secondLine: boolean) {
...
         changes = (
             <OverlayTrigger
                 key='changesRequestedDot'
                 placement='bottom'
-                overlay={<Tooltip id='changesRequestedTooltip'>{'Changes Requested'}</Tooltip>}
+                overlay={<Tooltip
+                    id='changesRequestedTooltip'
+                    aria-label={changedRequestedText}
+                // eslint-disable-next-line react/jsx-closing-bracket-location
+                >{'Changes Requested'}</Tooltip>}
```
Comment (mickmister) [inline webapp/src/components/sidebar_right/github_items.tsx:457]:
```suggestion
                overlay={
                    <Tooltip
                        id='changesRequestedTooltip'
                        aria-label={changedRequestedText}
                    >
                        {'Changes Requested'}
                    </Tooltip>
                }
```

File: webapp/src/components/sidebar_right/github_items.tsx [hunk 8]
```diff
@@ -381,11 +443,18 @@ function getReviewText(item: Item, style: any, secondLine: boolean) {
     }
 
     if (changesRequested > 0) {
+        // there are two conflicting rules so decided to choose one over the other on line 456
```
Comment (mickmister) [inline webapp/src/components/sidebar_right/github_items.tsx:446]: Please remove this comment. Plus the line numbers will change over time, so this will eventually be referencing the wrong line. Also, this should probably be a PR comment, not a comment in the code.

File: webapp/src/components/sidebar_right/github_items.tsx (hunk 3 above)
Comment (mickmister) [reply webapp/src/components/sidebar_right/github_items.tsx:173]: Looks like we should have the milestone's name here, based on the discussion in this thread

**Comment by mickmister** [comment]
> @mickmister this was an original pr by @sanjaydemansol I am not sure if it was ticked or not. Should I just send a new PR in place of this ? 

Since there is already review and dicussion history on this PR, I think it makes sense to finish this one that's in progress

**Review by mickmister — CHANGES_REQUESTED** [review]

File: webapp/src/components/sidebar_right/github_items.tsx (hunk 8 above)
Comment (maisnamrajusingh) [reply webapp/src/components/sidebar_right/github_items.tsx]: forgot to delete this one.

File: webapp/src/components/sidebar_right/github_items.tsx (hunk 4 above)
Comment (mickmister) [reply webapp/src/components/sidebar_right/github_items.tsx:198]: I think "Checks Passing" would be more appropriate, but not a required change

**Review by mickmister — APPROVED** [review]

**Review by DHaussermann — APPROVED** [review]
Tested and passed
- For the area in scope here:
  - All title spans have aria labels
  - Other spans and divs seem to all have aria label
  - All svg images have aria labels
- No regressions found
 
**note** The initial issue makes mention that some hover effects are missing. While all .svg now have a label this did not solve the problem of missing hover effects on some elements. However, if this is problematic we should open a new issue with specifically where we want text to show on hover. Therefore I believe we should accept this PR as is.   

LGTM!
